File: server/src/git_service.js
const { exec } = require('child_process');
const path = require('path');

const WORKSPACE_ROOT = process.cwd();

/**
 * Git Operations Service for Autonomous AI Coding Agents
 */
class GitService {
  constructor() {
    this.cwd = WORKSPACE_ROOT;
  }

  execGit(args) {
    return new Promise((resolve) => {
      exec(`git ${args}`, { cwd: this.cwd, timeout: 15000 }, (error, stdout, stderr) => {
        if (error) {
          return resolve({ success: false, error: error.message, stderr: stderr.trim() });
        }
        resolve({ success: true, stdout: stdout.trim() });
      });
    });
  }

  async getStatus() {
    const res = await this.execGit('status --porcelain');
    if (!res.success) return { isGitRepo: false, status: [] };

    const lines = res.stdout ? res.stdout.split('\n') : [];
    const files = lines.map(line => ({
      state: line.slice(0, 2).trim(),
      file: line.slice(3).trim()
    }));

    return { isGitRepo: true, files };
  }

  async getCurrentBranch() {
    const res = await this.execGit('rev-parse --abbrev-ref HEAD');
    return res.success ? res.stdout : 'main';
  }

  async createAgentBranch(agentName, featureName) {
    const branchName = `agent/${agentName.toLowerCase().replace(/\s+/g, '-')}-${featureName.toLowerCase().replace(/\s+/g, '-')}`;
    const res = await this.execGit(`checkout -b ${branchName}`);
    return { success: res.success, branch: branchName, message: res.stdout || res.error };
  }

  async commitAgentChanges(agentName, message) {
    await this.execGit('add .');
    const commitMsg = `[Agent: ${agentName}] ${message}`;
    const res = await this.execGit(`commit -m "${commitMsg.replace(/"/g, '\\"')}"`);
    return { success: res.success, message: res.stdout || res.error };
  }

  async getDiff() {
    const res = await this.execGit('diff');
    return res.success ? res.stdout : '';
  }
}

module.exports = new GitService();

File: python_backend/git_service.py
import subprocess
import os
from typing import Dict, Any

WORKSPACE_ROOT = os.getcwd()

class GitService:
    def __init__(self):
        self.cwd = WORKSPACE_ROOT

    def _exec_git(self, args: str) -> Dict[str, Any]:
        try:
            res = subprocess.run(
                f"git {args}",
                shell=True,
                cwd=self.cwd,
                capture_output=True,
                text=True,
                timeout=15
            )
            return {"success": res.returncode == 0, "stdout": res.stdout.strip(), "stderr": res.stderr.strip()}
        except Exception as e:
            return {"success": False, "error": str(e)}

    def get_status(self) -> Dict[str, Any]:
        res = self._exec_git("status --porcelain")
        if not res["success"]:
            return {"isGitRepo": False, "status": []}

        lines = res["stdout"].split("\n") if res["stdout"] else []
        files = [{"state": line[:2].strip(), "file": line[3:].strip()} for line in lines if len(line) >= 4]
        return {"isGitRepo": True, "files": files}

    def create_branch(self, agent_name: str, feature_name: str) -> Dict[str, Any]:
        clean_agent = agent_name.lower().replace(" ", "-")
        clean_feat = feature_name.lower().replace(" ", "-")
        branch_name = f"agent/{clean_agent}-{clean_feat}"
        res = self._exec_git(f"checkout -b {branch_name}")
        return {"success": res["success"], "branch": branch_name, "message": res.get("stdout") or res.get("stderr")}

    def commit_changes(self, agent_name: str, message: str) -> Dict[str, Any]:
        self._exec_git("add .")
        clean_msg = message.replace('"', '\\"')
        res = self._exec_git(f'commit -m "[Agent: {agent_name}] {clean_msg}"')
        return {"success": res["success"], "message": res.get("stdout") or res.get("stderr")}

git_service = GitService()
